ingestion/src/test/java/feast/ingestion/ImportJobTest.java

```diff
@@ -180,91 +185,332 @@ public void runPipeline_ShouldWriteToRedisCorrectlyGivenValidSpecAndFeatureRow()
+    publishFeatureRows(input, pipelineResult);
+
+    LOGGER.info("Validating the actual values written to Redis ...");
+    StatefulRedisConnection<byte[], byte[]> connection = redisClient.connect(new ByteArrayCodec());
+    RedisCommands<byte[], byte[]> sync = connection.sync();
+    expected.forEach(
+        (key, expectedValue) -> {
+
+          // Ensure ingested key exists.
+          byte[] actualByteValue = getBytesFromKey(sync, key);
+          if (actualByteValue == null) {
+            Assert.fail("Missing key in Redis.");
           }
 
+          // Ensure value is a valid serialized FeatureRow object.
+          FeatureRow actualValue = getFeatureRowFromByte(key, actualByteValue);
+
           // Ensure the retrieved FeatureRow is equal to the ingested FeatureRow.
           Assert.assertEquals(expectedValue, actualValue);
         });
     redisClient.shutdown();
   }
+
+  // get the expected rows from the input, and default values.
+  private Map<RedisKey, FeatureRow> getExpectedRows(List<FeatureRow> input, FeatureSetSpec spec) {
+    Map<RedisKey, FeatureRow> expected = new HashMap<>();
+    input.forEach(
+        i -> {
+          if (spec.getEntitiesList().stream()
+              .allMatch(
+                  entity ->
+                      i.getFieldsList().stream()
+                          .map(Field::getName)
+                          .collect(Collectors.toSet())
+                          .contains(entity.getName()))) {
+            RedisKey redisKey = TestUtil.createRedisKey(spec, i);
+            FeatureRow value = getFieldsFromFeatureRow(spec, i);
+            expected.put(redisKey, value);
+          }
+        });
+    return expected;
+  }
+
+  private byte[] getBytesFromKey(RedisCommands<byte[], byte[]> sync, RedisKey key) {
+    byte[] actualByteValue = sync.get(key.toByteArray());
+    if (actualByteValue == null) {
+      LOGGER.error("Key not found in Redis: " + key);
+      LOGGER.info("Redis INFO:");
+      LOGGER.info(sync.info());
+      byte[] randomKey = sync.randomkey();
+      if (randomKey != null) {
+        LOGGER.info("Sample random key, value (for debugging purpose):");
+        LOGGER.info("Key: " + randomKey);
+        LOGGER.info("Value: " + sync.get(randomKey));
+      }
+    }
+    return actualByteValue;
+  }
+
+  private FeatureRow getFeatureRowFromByte(RedisKey key, byte[] actualByteValue) {
+    FeatureRow actualValue = null;
+    try {
+      actualValue = FeatureRow.parseFrom(actualByteValue);
+    } catch (InvalidProtocolBufferException e) {
+      Assert.fail(
+          String.format(
+              "Actual Redis value cannot be parsed as FeatureRow, key: %s, value :%s",
+              key, new String(actualByteValue, StandardCharsets.UTF_8)));
+    }
+    return actualValue;
+  }
+
+  // Convert feature rows to a list of values for each non-entity feature.
+  private FeatureRow getFieldsFromFeatureRow(FeatureSetSpec spec, FeatureRow row) {
```

Comment:
Can we rename and update comments for this method so the name and comments are more representative of what this method is doing?

e.g.
- name is `getFields...` but it actually returns `FeatureRow`
- `spec` is passed in as parameter, but it's not clear without looking at the implementation details how this spec argument is used